pkg/sql/catalog/descs/hydrate.go

```diff
@@ -12,369 +12,165 @@ package descs
-	// Collect the needed information to set up metadata in those types.
-	dbDescs := make(map[descpb.ID]catalog.DatabaseDescriptor)
-	typDescs := make(map[descpb.ID]catalog.TypeDescriptor)
-	schemaDescs := make(map[descpb.ID]catalog.SchemaDescriptor)
-	for _, desc := range descs {
-		switch desc := desc.(type) {
-		case catalog.DatabaseDescriptor:
-			dbDescs[desc.GetID()] = desc
-		case catalog.TypeDescriptor:
-			typDescs[desc.GetID()] = desc
-		case catalog.SchemaDescriptor:
-			schemaDescs[desc.GetID()] = desc
-		}
-	}
-	// If we found any type descriptors, that means that some of the tables we
-	// scanned might have types that need hydrating.
-	if len(typDescs) > 0 {
-		// Since we just scanned all the descriptors, we already have everything
-		// we need to hydrate our types. Set up an accessor for the type hydration
-		// method to look into the scanned set of descriptors.
-		typeLookup := func(ctx context.Context, id descpb.ID) (tree.TypeName, catalog.TypeDescriptor, error) {
-			typDesc, ok := typDescs[id]
-			if !ok {
-				n := tree.MakeUnresolvedName(fmt.Sprintf("[%d]", id))
-				return tree.TypeName{}, nil, sqlerrors.NewUndefinedObjectError(&n,
-					tree.TypeObject)
-			}
-			dbDesc, ok := dbDescs[typDesc.GetParentID()]
-			if !ok {
-				n := fmt.Sprintf("[%d]", typDesc.GetParentID())
-				return tree.TypeName{}, nil, sqlerrors.NewUndefinedDatabaseError(n)
-			}
-			// We don't use the collection's ResolveSchemaByID method here because
-			// we already have all of the descriptors. User defined types are only
-			// members of the public schema or a user defined schema, so those are
-			// the only cases we have to consider here.
-			var scName string
-			switch typDesc.GetParentSchemaID() {
-			// TODO(richardjcai): Remove case for keys.PublicSchemaID in 22.2.
-			case keys.PublicSchemaID:
-				scName = tree.PublicSchema
-			default:
-				scName = schemaDescs[typDesc.GetParentSchemaID()].GetName()
-			}
-			name := tree.MakeQualifiedTypeName(dbDesc.GetName(), scName, typDesc.GetName())
-			return name, typDesc, nil
-		}
-		// Now hydrate all table descriptors.
-		for i := range descs {
+	// Replace immutable hydratable descriptors in the slice with hydrated copies
+	// from the cache, or otherwise by creating a copy and hydrating it.
+	if len(hydratableImmutableIndexes) > 0 {
+		typeFn := makeImmutableTypeLookupFunc(tc, txn, flags, descs)
+		for _, i := range hydratableImmutableIndexes {
 			desc := descs[i]
-			// Never hydrate dropped descriptors.
-			if desc.Dropped() {
-				continue
+			// Utilize the cache of hydrated tables if we have one and this descriptor
+			// was leased.
+			// TODO(ajwerner): Consider surfacing the mechanism used to retrieve the
+			// descriptor up to this layer.
+			if hd := desc.(catalog.HydratableDescriptor); tc.canUseHydratedDescriptorCache(hd.GetID()) {
+				if cached, err := tc.hydrated.GetHydratedDescriptor(ctx, hd, typeFn); err != nil {
```

Comment:
I think we're in a position now where we can improve this cache a bit. Possibly get rid of it entirely!